### PR TITLE
Refactor agent interface a bit

### PR DESCRIPTION
**Pull request by rbren**
Right now,  the LangChain agent also includes all its own logic for running commands, browsing, reading/writing files, etc. That needs to get factored out so that all agents can use it.

From what I can tell from the recent refactor, we basically just instantiate an agent with a task, and call run(), then it continues until it finishes.

This refactors things so that we have a flow more like this:

* server receives a task request
* server instantiates an agent
* server starts a control loop:
  * in each iteration, call agent.step()
  * agent.step() is allowed to send at most 1 LLM prompt
  * agent.step() returns an action, one of:
    * read file
    * write file
    * run cmd
    * browse
    * no-op
  * looper executes the action
    * and optionally passes messages on to e.g. the client browser

### CODE REVIEW

**Review by neubig — APPROVED** [review]
This looks great to me! I also looped in @xingyaoww in case he wants to take a look before we merge this in.

File: .gitignore
Comment (rbren) [inline .gitignore:17]: This is probably not what was intended here...but also `lib` is a typical code directory. Is there something more pythonic I can call this dir? `pkg`?

File: opendevin/agent.py
Comment (rbren) [inline opendevin/agent.py:74]: `add_event(Event('chat', {'message': 'foo'}))` takes the place of `chat`

File: opendevin/lib/event.py
Comment (rbren) [reply opendevin/lib/event.py:5]: Hmm this is a really good point.

I do think we'll want structured actions outside the agent to highlight certain things in the UI (e.g. if we know which file was written, we can change what file is visible in the UI editor, or if we know it's browsing, we can highlight the browser). 

But letting the agent specify the natural language message is also important. Maybe we add a `message` field to `Event`?

Worth noting: with the current setup, not every `Event` is created by the agent. Some Events represent command output, file contents, or HTML from the browser. So not all of them would have a `message`--but we could display the ones that do in the chat log.

Comment (xingyaoww) [reply opendevin/lib/event.py:5]: Good point! We will be needing some metadata for the UI -- If that's the case, maybe we can make this into an Action class with two types: Message or Action (and under action, we can browse, read, write, etc).

But i guess the tricky thing is that those changes would be pretty hard to detect, if the model modifies those file with `bash` instead of using the provided tool. but let's worry about it when it comes up

Comment (xingyaoww) [reply opendevin/lib/event.py:5]: Maybe we can just do 

The input to Agent:
- natural language 

The output of Agent:
- Action: which is Message OR Event

then we have 
- class Message (as is)
- class Event -> that encapsulates most action you have now?

File: opendevin/agent.py
Comment (rbren) [reply opendevin/agent.py:102]: So there are two things I'm trying to control here:
* `add_event` is meant to keep a history of `Events`, which should include files getting read, commands getting run, log outputs, browse actions, etc. I think we need _structured_ data about all that happening, e.g. to drive the UI. I think that history should be kept in the abstract interface or in the controller, but the underlying agent gets a chance to react to each event. Maybe `add_event` isn't the right name--maybe something like `event_callback` or `on_event`
* `step` is just meant to move the agent forward one step. I don't love the way I'm plumbing `CommandManger` through--it's a little messy and doesn't separate concerns properly. It's only there so the agent can report on what's running in the background. It'd be better to make this a `state` object on the agent (which might include other things, like git status and cwd)

File: opendevin/controller.py
Comment (rbren) [reply opendevin/controller.py:31]: Do you mean the truncation logic specifically? I agree it'd be better to leave that to the agent--I will factor that part out.

But for the actions themselves, yes I think it's better to give the agent a "menu" of possible actions, and leave action execution outside of the agent. For two reasons:
* it simplifies the agent's contract--the agent just has to translate natural language into a stream of Events
* it makes things predictable--the UI can reflect _exactly_ what the agent is doing at any given time. IMO this part is super important: the user is going to want a high degree of visibility into how the AI is behaving. E.g. if there's an `error` event, we show it in red, if there's a `browse` event we bring the browser in focus, etc.

Comment (xingyaoww) [reply opendevin/controller.py:31]: For UI purposes, how about we expect each agent to generate an `Event` (when each action that need to be tracked by UI is triggered - as discussed above) but do not assume how these `Event` will be generated by the `Agent`?

I worry about the fixed action space is because CodeAct agent's action space is just bash commands - if we make it fixed, we might have trouble support different agent variants.

File: .gitignore
Comment (xingyaoww) [reply .gitignore:17]: i don't know maybe `utils`?

**Pull request merged**
